04-linear-chain.Rmd

```diff
@@ -0,0 +1,526 @@
+
+```         
+task MarkDuplicates {
+  input {
+    File input_bam
+  }
+  String base_file_name = basename(input_bam, ".sorted_query_aligned.bam")
+}
+```
+
+### Command
+
+As we have been doing all along, we refer to any defined variables from the input or private variables using \~{this} syntax.
+
+```         
+task MarkDuplicates {
+  input {
+    File input_bam
+  }
+
+  String base_file_name = basename(input_bam, ".sorted_query_aligned.bam")
+
+  command <<<
+    gatk MarkDuplicates \
+      --INPUT "~{input_bam}" \
+      --OUTPUT "~{base_file_name}.duplicates_marked.bam" \
+      --METRICS_FILE "~{base_file_name}.duplicate_metrics" \
+      --CREATE_INDEX true \
+      --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 \
+      --VALIDATION_STRINGENCY SILENT
+  >>>
+}
+```
+
+### Runtime and Output
+
+We specify a different Docker image that contains the GATK software, and the relevant computing needs. We also specify three different output files, two of which are specified in the command section, and the third is a bam index file that is automatically created by the command section.
+
+Below is the task all together. It has a form very similar to our first task `BwaMem`.
+
+```         
+ task MarkDuplicates {
+  input {
+    File input_bam
+  }
+
+  String base_file_name = basename(input_bam, ".sorted_query_aligned.bam")
+
+  command <<<
+    gatk MarkDuplicates \
+      --INPUT "~{input_bam}" \
+      --OUTPUT "~{base_file_name}.duplicates_marked.bam" \
+      --METRICS_FILE "~{base_file_name}.duplicate_metrics" \
+      --CREATE_INDEX true \
+      --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 \
+      --VALIDATION_STRINGENCY SILENT
+  >>>
+  
+  runtime {
+    docker: "broadinstitute/gatk:4.1.4.0"
+    memory: "48 GB"
+    cpu: 4
+  }
+  
+  output {
+    File markDuplicates_bam = "~{base_file_name}.duplicates_marked.bam"
+    File markDuplicates_bai = "~{base_file_name}.duplicates_marked.bai"
+    File duplicate_metrics = "~{base_file_name}.duplicate_metrics"
+  }
+}
+```
+
+### Testing the workflow
+
+As before, when you add a new task to the workflow, you should always test that it works on your test sample. To check that `MarkDuplicates` is indeed marking PCR duplicates, you could check for the presence of the [PCR duplicate flag in reads](https://gatk.broadinstitute.org/hc/en-us/articles/360051306171-MarkDuplicates-Picard), which has a decimal value of 1024 in the SAM Flags Field.
+
+## The rest of the linear chain workflow
+
+We build out the rest of the tasks in a very similar fashion. Tasks `ApplyBaseRecalibrator` and `Mutect2TumorOnly` both have files that need to be localized, but otherwise all the tasks have a very similar form as `BwaMem` and `MarkDuplicates`. For somatic mutation calling in the `Mutect2TumorOnly` task, we currently only use the tumor sample. Our complete workflow will require a pair of tumor and normal samples for each individual.
```

Comment:
This section introduces the idea of having a tumor sample and a normal sample each running versions of this task, but since task aliasing will be introduced later, it's a little bit confusing to bring it up here.

I'd recommend not bringing up the normal sample at all until we get to that chapter in particular.

Reply:
Noted. Though I think I need to bring it up a little bit because our workflow plan talks about tumor and normal samples, and it would also be confusing to see it not applied here. I will try to word it more clearly